File: kotlin-multiplatform/monstradore/androidApp/src/main/java/com/example/monstradore/android/uiux/UIElementsView.kt
package com.example.monstradore.android.uiux

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Home
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.example.monstradore.android.R
import kotlinx.coroutines.launch

@Composable
fun UIElementsContent(navController: NavController) {
    Column(
        modifier = Modifier
            .padding(5.dp)
            .verticalScroll(rememberScrollState())
    ) {
        Text(
            text = "Grundlegende Elemente",
            fontSize = 25.sp,
            fontWeight = FontWeight.Bold
        )
        BasicElements()
        Text(
            text = "Elemente mit Statusverwaltung",
            fontSize = 25.sp,
            fontWeight = FontWeight.Bold
        )
        StateElements()
        Text(
            text = "Plattformspezifische Elemente",
            fontSize = 25.sp,
            fontWeight = FontWeight.Bold
        )
        SpecificNavigation(navController)
        Text(
            text = "Fortgeschrittene Elemente",
            fontSize = 25.sp,
            fontWeight = FontWeight.Bold
        )
        ActionElements()
    }
}

@Composable
fun BasicElements() {
    Column {
        Text(
            text = "Text",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Text(text = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.")
        Text(
            text = "Bild",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Image(
            painter = painterResource(id = R.drawable.sample),
            contentDescription = stringResource(id = R.string.sample_img)
        )
        Text(
            text = "Liste",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Column {
            repeat(5) {
                Text(text = "Item: $it")
            }
        }
        Text(
            text = "Button",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Button(onClick = { }) {
            Text(text = "Button")
        }
        Text(
            text = "Icon Button",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        IconButton(onClick = { }) {
            Icon(Icons.Outlined.Home, contentDescription = "Icon Button description")
        }
    }
}

@Composable
fun StateElements() {
    var sliderValue by remember { mutableStateOf(0f) }
    var text by remember { mutableStateOf("") }
    val checkedState = remember { mutableStateOf(true) }
    Column {
        Text(
            text = "Slider",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Slider(value = sliderValue, onValueChange = { newValue ->
            sliderValue = newValue
        }
        )
        Text(
            text = "Textfeld",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        TextField(
            value = text,
            onValueChange = { text = it },
            label = { Text("Label") }
        )
        Text(
            text = "Switch / Toggle",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold
        )
        Switch(
            checked = checkedState.value,
            onCheckedChange = { checkedState.value = it }
        )
    }
}

@Composable
fun ActionElements() {
    Row(
        horizontalArrangement = Arrangement.spacedBy(5.dp)
    ) {
        MenuElement()
        DialogElement()
        //BottomSheetElement()
    }
}

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun BottomSheetElement() {
    val scope = rememberCoroutineScope()
    val scaffoldState = rememberBottomSheetScaffoldState(
        bottomSheetState = BottomSheetState(BottomSheetValue.Collapsed)
    )
    BottomSheetScaffold(
        sheetContent = {
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(500.dp),
                contentAlignment = Alignment.Center
            ) {
                Text("Text in Bottom Sheet")
            }
        },
        scaffoldState = scaffoldState,
        sheetPeekHeight = 0.dp
    ) {
        Button(onClick = {
            scope.launch {
                if (scaffoldState.bottomSheetState.isCollapsed) {
                    scaffoldState.bottomSheetState.expand()
                } else {
                    scaffoldState.bottomSheetState.collapse()
                }
            }
        }) {
            Text(text = "Bottom Sheet")
        }
    }
}

@Composable
fun MenuElement() {
    var showMenu by remember { mutableStateOf(false) }
    Box {
        Button(onClick = { showMenu = !showMenu }) {
            Text(text = "Menu")
        }
        DropdownMenu(
            expanded = showMenu,
            onDismissRequest = { showMenu = false }) {
            DropdownMenuItem(onClick = { }) {
                Text(text = "Menu Item")
            }
        }
    }
}

@Composable
fun DialogElement() {
    Column {
        val openDialog = remember { mutableStateOf(false) }
        Button(onClick = {
            openDialog.value = true
        }) {
            Text("Dialog")
        }
        if (openDialog.value) {
            AlertDialog(
                onDismissRequest = {
                    openDialog.value = false
                },
                title = {
                    Text(text = "Dialog Titel")
                },
                text = {
                    Text("Dialog Text")
                },
                confirmButton = {
                    TextButton(
                        onClick = {
                            openDialog.value = false
                        }) {
                        Text("OK")
                    }

                },
                dismissButton = {
                    TextButton(
                        onClick = {
                            openDialog.value = false
                        }) {
                        Text("Cancel")
                    }
                }
            )
        }
    }
}

@Composable
fun SpecificNavigation(navController: NavController) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(5.dp)
    ) {
        Button(onClick = { navController.navigate("androidelements") }) {
            Text(text = "Android")
        }
        Button(onClick = { navController.navigate("ioselements") }) {
            Text(text = "iOS")
        }
    }
}

File: kotlin-multiplatform/monstradore/iosApp/iosApp/uiux/UIElementsView.swift
//
//  UIElementsView.swift
//  iosApp
//
//  Created by Anja on 19.11.22.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI

struct UIElementsView: View {
    var body: some View {
        NavigationView {
            ScrollView {
                VStack(alignment: .leading)
                {
                    BasicElements()
                    ButtonElements()
                    StateElements()
                    PlatformspecificElements()
                    AdvancedElements()
                }
                .padding()
                .navigationBarTitleDisplayMode(.inline)
            }
        }
    }
}

//struct UIElementsView_Previews: PreviewProvider {
//    static var previews: some View {
//        UIElementsView()
//    }
//}

struct BasicElements: View {
    var body: some View {
        Text("Grundlegende Elemente")
            .font(.title)
            .multilineTextAlignment(.leading)
        Text("Text")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Text("Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.")
            .multilineTextAlignment(.leading)
        Text("Bild")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Image("sample")
            .resizable()
            .aspectRatio(contentMode: .fit)
        Text("Liste")
            .font(.headline)
            .multilineTextAlignment(.leading)
        List {
            Text("Item: 0")
            Text("Item: 1")
            Text("Item: 2")
            Text("Item: 3")
            Text("Item: 4")
        }
        .frame(height: 280)
    }
}

struct ButtonElements: View {
    var body: some View {
        Text("Button")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Button(action: {}) {
            Text("Button")
        }
        Text("Icon Button")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Button { } label: {
            Image(systemName: "house")
        }
    }
}

struct StateElements: View {
    @State private var speed = 0.0
    @State private var isEditing = false
    @State private var textInput = ""
    @State private var toggleState = true
    var body: some View {
        Text("Elemente mit Statusverwaltung")
            .font(.title)
            .multilineTextAlignment(.leading)
        Text("Slider")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Slider(
            value: $speed,
            in: 0...100,
            onEditingChanged: { editing in
                isEditing = editing
            }
        )
        Text("Textfeld")
            .font(.headline)
            .multilineTextAlignment(.leading)
        TextField("Texteingabe", text: $textInput)
        Text("Switch / Toggle")
            .font(.headline)
            .multilineTextAlignment(.leading)
        Toggle("", isOn: $toggleState)
    }
}

struct PlatformspecificElements: View {
    @State private var showView: Int? = 0
    var body: some View {
        Text("Plattformspezifische Elemente")
            .font(.title)
            .multilineTextAlignment(.leading)
            HStack {
                NavigationLink(destination: AndroidElementsView(), tag: 1, selection: $showView) {
                    EmptyView()
                }
                NavigationLink(destination: iOSElementsView(), tag: 2, selection: $showView) {
                    EmptyView()
                }
                Button("Android", action: {
                    self.showView = 1
                })
                Button("iOS", action: {
                    self.showView = 2
                })
        }
    }
}

struct AdvancedElements: View {
    @State private var showAlert = false
    var body: some View {
        Text("Fortgeschrittene Elemente")
            .font(.title)
            .multilineTextAlignment(.leading)
        HStack {
            Button(action: { }) {
                Text("Menu")
            }
            .contextMenu {
                Button { } label: {
                    Label("Menu Item", systemImage: "")
                }
            }
            Button("Dialog") {
                showAlert = true
            }
            .alert(isPresented: $showAlert) {
                Alert(
                    title: Text("Dialog Titel"),
                    message: Text("Dialog Text"),
                    dismissButton: .default(Text("OK"))
                )
            }
        }
    }
}
